File: week_1/code/remove_nth_node.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        """
        Given the head of a linked list, remove the nth node from the end of the list and return its head.
        """
        dummy = ListNode(0, head)
        left = dummy
        right = head
        
        while n > 0 and right:
            right = right.next
            n -= 1
            
        while right:
            left = left.next
            right = right.next
            
        # Delete the node
        left.next = left.next.next
        
        return dummy.next

# Helper to print list
def printList(head):
    vals = []
    curr = head
    while curr:
        vals.append(str(curr.val))
        curr = curr.next
    print(" -> ".join(vals))

# Helper to create list
def createList(arr):
    if not arr: return None
    head = ListNode(arr[0])
    curr = head
    for val in arr[1:]:
        curr.next = ListNode(val)
        curr = curr.next
    return head

# Test cases
if __name__ == "__main__":
    sol = Solution()
    
    # Test 1: 1->2->3->4->5, n=2
    head1 = createList([1,2,3,4,5])
    print("Original 1:")
    printList(head1)
    new_head1 = sol.removeNthFromEnd(head1, 2)
    print("Removed 2nd from end (Expected: 1->2->3->5):")
    printList(new_head1)

    # Test 2: 1, n=1
    head2 = createList([1])
    print("Original 2:")
    printList(head2)
    new_head2 = sol.removeNthFromEnd(head2, 1)
    print("Removed 1st from end (Expected: Empty):")
    printList(new_head2)

    # Test 3: 1->2, n=1
    head3 = createList([1,2])
    print("Original 3:")
    printList(head3)
    new_head3 = sol.removeNthFromEnd(head3, 1)
    print("Removed 1st from end (Expected: 1):")
    printList(new_head3)

File: week_1/code/remove_nth_node.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
}

/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function (head, n) {
    let dummy = new ListNode(0, head);
    let left = dummy;
    let right = head;

    while (n > 0 && right) {
        right = right.next;
        n -= 1;
    }

    while (right) {
        left = left.next;
        right = right.next;
    }

    // Delete the node
    left.next = left.next.next;

    return dummy.next;
};

// Helper to print list
function printList(head) {
    let vals = [];
    let curr = head;
    while (curr) {
        vals.push(curr.val);
        curr = curr.next;
    }
    console.log(vals.join(" -> "));
}

// Helper to create list
function createList(arr) {
    if (!arr.length) return null;
    let head = new ListNode(arr[0]);
    let curr = head;
    for (let i = 1; i < arr.length; i++) {
        curr.next = new ListNode(arr[i]);
        curr = curr.next;
    }
    return head;
}

// Test cases
// Test 1: 1->2->3->4->5, n=2
let head1 = createList([1, 2, 3, 4, 5]);
console.log("Original 1:");
printList(head1);
let newHead1 = removeNthFromEnd(head1, 2);
console.log("Removed 2nd from end (Expected: 1->2->3->5):");
printList(newHead1);

// Test 2: 1, n=1
let head2 = createList([1]);
console.log("Original 2:");
printList(head2);
let newHead2 = removeNthFromEnd(head2, 1);
console.log("Removed 1st from end (Expected: Empty):");
printList(newHead2);

// Test 3: 1->2, n=1
let head3 = createList([1, 2]);
console.log("Original 3:");
printList(head3);
let newHead3 = removeNthFromEnd(head3, 1);
console.log("Removed 1st from end (Expected: 1):");
printList(newHead3);
